Make agent channel setup lazy.

fspmarshall opened the pull request:
Changes agent channel setup behavior to be consistent with openssh by having servers lazily request agent channels when they are needed, rather than immediately starting a single connection-wide channel as soon as forwarding is requested.

Fixes an issue introduced in #3613 which caused openssh clients to hang on exit due to persistent agent channel.

webvictim commented on lib/sshutils/ctx.go:
```diff
@@ -42,24 +46,57 @@ type ConnectionContext struct {
+	// closed indicates that closers have been run.
+	closed bool
+
+	// cancel cancels the context.Context scope assocaited with this ConnectionContext.
```
associated

webvictim commented on lib/teleagent/lazy.go:
```diff
@@ -0,0 +1,112 @@
+
+// lazyAgent is an agent.Agent instance which is lazily initialized
+// via a call to an AgentGetter.  Helpful for dealing with APIs
+// which require, but may not acutally use, an agent.
```
actually

awly commented on lib/sshutils/ctx.go:
```diff
@@ -29,6 +32,7 @@ import (
 
 // ConnectionContext manages connection-level state.
 type ConnectionContext struct {
+	context.Context
```
Don't embed the context, pass it via func arguments instead: https://github.com/golang/go/wiki/CodeReviewComments#contexts
For premature cancellation, the caller can wrap using `context.WithCancel`.
For regular cleanup, consider adding a `Close` method where necessary

fspmarshall replied:
I've removed the other cases of contexts in structs from this PR, but this (and `srv.ServerContext`) is a bit different.  My thinking about embedding in these cases is that these already were basically `Context` values that just happened to not implement the full interface.  They already where literally request-scoped parameters which carry some request-level state and also assist in signalling cancellation (they even have a `Context`-esque parent-child relationship).  I decided to embed `context.Context` here because it was simpler and easier to just make it official.

awly commented on lib/sshutils/ctx.go:
```diff
@@ -88,37 +125,44 @@ func (c *ConnectionContext) ExportEnv(m map[string]string) {
+	// if context was already closed, run the closer immediately
+	// in the background.
+	if c.closed {
+		go func() {
+			closer.Close()
+		}()
```
this can be: `go closer.Close()`
but why in a goroutine?

fspmarshall replied:
Calling `AddCloser` in a scenario where `ConnectionContext` has already started closing should be a fairly rare event.  As such, I figure its better from a debugging perspective to be able to assume that closers are always run asynchronously, rather than having a small chance of the method blocking on the closer.

awly commented on lib/sshutils/server.go:
```diff
@@ -432,7 +432,11 @@ func (s *Server) HandleConnection(conn net.Conn) {
+	// closeContext field is used to trigger starvation on cancellation by halting
+	// the acceptance of new connections; it is not intended to halt in-progress
+	// connection handling, and is therefore orthogonal to the role of ConnectionContext.
+	ccx := NewConnectionContext(context.TODO(), wconn, sconn)
```
use `context.Background()`

awly commented on lib/teleagent/agent.go:
```diff
@@ -26,6 +31,33 @@ func NewServer() *AgentServer {
 	return &AgentServer{Agent: agent.NewKeyring()}
```
should this always take an `AgentGetter` argument?

awly commented on lib/teleagent/agent.go:
```diff
@@ -26,6 +31,33 @@ func NewServer() *AgentServer {
+		defer cancel()
+		if err := agent.ServeAgent(instance, conn); err != nil {
+			if err != io.EOF {
+				log.Errorf(err.Error())
```
`log.Error(err)`

awly commented on lib/teleagent/agent.go:
```diff
@@ -47,9 +79,14 @@ func (a *AgentServer) ListenUnixSocket(path string, uid, gid int, mode os.FileMo
 
 // Serve starts serving on the listener, assumes that Listen was called before
 func (a *AgentServer) Serve() error {
+	ctx, cancel := context.WithCancel(context.TODO())
```
use `context.Background()` or pass a context from the caller

awly commented on lib/teleagent/lazy.go:
```diff
@@ -0,0 +1,112 @@
+// via a call to an AgentGetter.  Helpful for dealing with APIs
+// which require, but may not actually use, an agent.
+type lazyAgent struct {
+	ctx  context.Context
```
same as above: https://github.com/golang/go/wiki/CodeReviewComments#contexts
pass context through func arguments instead of struct fields

awly commented on lib/teleagent/lazy.go:
```diff
@@ -0,0 +1,112 @@
+
+// Lazy wraps an AgentGetter in an agent.Agent interface which lazily
+// initializes the agent on first method invocation.
+func Lazy(ctx context.Context, get AgentGetter) agent.Agent {
```
https://github.com/golang/go/wiki/CodeReviewComments#interfaces return a specific exported type `*LazyAgent` (or `*Lazy` and rename this func to `NewLazy`)

awly commented on lib/sshutils/ctx.go:
```diff
@@ -42,24 +46,57 @@ type ConnectionContext struct {
+// StartAgentChannel sets up a new agent forwarding channel against this connection.  The channel
+// is automatically closed when either ConnectionContext, or the supplied context.Context
+// gets canceled.
+func (c *ConnectionContext) StartAgentChannel(ctx context.Context) (agent.Agent, error) {
```
All the context plumbing here is added just to clean up the channel.
How about wrapping `agent.Agent` with a new type:
```
type ForwardedAgent struct {
  agent.Agent
  agentChannel io.Closer
}

func (a *ForwardedAgent) Close() error {
  return a.agentChannel.Close()
}
```

Handling the context cancellation from higher levels is still nice, but the caller can deal with it by calling Close in a defer.

russjones commented on lib/srv/regular/sshserver.go:
```diff
@@ -765,8 +764,9 @@ func (s *Server) serveAgent(ctx *srv.ServerContext) error {
-	agentServer := &teleagent.AgentServer{Agent: ctx.Parent.GetAgent()}
+	// start an agent server on a unix socket.  each incoming connection
+	// will result in a separate agent request.
+	agentServer := &teleagent.AgentServer{Getter: teleagent.AgentGetter(ctx.Parent.StartAgentChannel)}
```
I know you just copied over code, but we should use a function here, like `teleagent.New(...)`. It's not clear which/if the other fields need to be set.

russjones commented on lib/srv/regular/sshserver.go:
```diff
@@ -1174,14 +1174,7 @@ func (s *Server) handleAgentForwardNode(req *ssh.Request, ctx *srv.ServerContext
-
-	// save the agent in the context so it can be used later
-	ctx.Parent.SetAgent(agent.NewClient(authChannel), authChannel)
+	ctx.Parent.SetForwardAgent(true)
```
I wonder if a better name would help clarify what was is happening here? If I understand correctly at this point we want to allocate a lazy agent that will open a channel when the first request comes through. So maybe `ctx.Parent.StartLazyAgent()`?

fspmarshall replied:
This method doesn't start anything.  It just sets a flag, indicating that the context is now *capable* of agent forwarding (i.e. the client has sent an agent forwarding request and the server has verified that the associated teleport user is allowed to use agent forwarding).  This flag existed implicitly when we used a single global `Agent` because it would be `nil` until the request had gone through.

A boolian flag named `ForwardAgent` seemed like a good way to represent this, but I'm open to alternatives... does `AllowAgentForwarding` get the point across better?

russjones commented on lib/sshutils/ctx.go:
```diff
@@ -88,37 +125,44 @@ func (c *ConnectionContext) ExportEnv(m map[string]string) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	return c.agent
+	if c.getForwardAgent() {
+		return teleagent.Lazy(c.Context, c.StartAgentChannel)
+	}
```
Can this logic be encapsulated within the new lazy agent?

fspmarshall commented:
retest this please

fspmarshall committed:
Make agent channel setup lazy.

Changes agent channel setup behavior to be consistent
openssh by having servers lazily request agent channels
when they are needed, rather than immediately starting a
single connection-wide channel as soon as forwarding is
requested.  Fixes an issue introduced in #3613 which
caused openssh clients to hang on exit due to persistent
agent channel.

awly commented on lib/srv/regular/proxy_test.go:
```diff
@@ -35,52 +36,54 @@ func (s *ProxyTestSuite) SetUpSuite(c *check.C) {
+			// never actually be empty.
+			t.namespace = defaults.Namespace
+		}
+		cmt := check.Commentf("Test case %d: %+v", i, t)
```
optional: you can add a `desc string` field to the test cases to explain what the case covers instead of printing index + contents

fspmarshall replied:
Cool, I wasn't aware of that feature.  Think I'll keep as-is though; since I'm mutating the test case I think its nice to include a pretty print of the test case's value.

awly commented on lib/srv/regular/sshserver.go:
```diff
@@ -1007,31 +1006,33 @@ func (s *Server) handleDirectTCPIPRequest(ccx *sshutils.ConnectionContext, ident
 	// session request is complete.
-	ctx, err := srv.NewServerContext(ccx, s, identityContext)
+
+	//nolint:ineffassign,staticcheck
```
why disable the linter here?

awly commented on lib/srv/regular/sshserver.go:
```diff
@@ -1328,18 +1317,20 @@ func (s *Server) handleVersionRequest(req *ssh.Request) {
 	// session request is complete.
-	ctx, err := srv.NewServerContext(ccx, s, identityContext)
+
+	//nolint:ineffassign,staticcheck
```
and also here

awly commented on lib/teleagent/agent.go:
```diff
@@ -14,16 +14,63 @@ import (
+// supplied agent.Agent instance has a Close method.  This means that
+// WrapAgent can also be called on an instance of the extended Agent
+// interface in order to protect its Close method from being called.
+func WrapAgent(std agent.Agent) Agent {
```
nit: how about naming this `NopCloser` to communicate the effects more clearly?

awly commented on lib/teleagent/agent.go:
```diff
@@ -14,16 +14,63 @@ import (
+}
+
+// AgentGetter is a function used to get an agent instance.
+type AgentGetter func() (Agent, error)
```
how about just `Getter` to make this `teleagent.Getter`?
https://github.com/golang/go/wiki/CodeReviewComments#package-names

awly commented on lib/teleagent/agent.go:
```diff
@@ -14,16 +14,63 @@ import (
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	go func() {
```
hmm, don't `Serve` calls generally block?
like `http.Server.Serve` for example

awly commented on lib/teleagent/agent.go:
```diff
@@ -14,16 +14,63 @@ import (
+		defer instance.Close()
+		if err := agent.ServeAgent(instance, conn); err != nil {
+			if err != io.EOF {
+				log.Error(err.Error())
```
nit: `log.Error(err)`

fspmarshall commented on lib/srv/forward/sshserver.go:
```diff
@@ -726,79 +720,80 @@ func (s *Server) handleDirectTCPIPRequest(ch ssh.Channel, req *sshutils.DirectTC
 	// done on the server's terminating side.
-	ctx, err := srv.NewServerContext(s.connectionContext, s, s.identityContext)
+
+	//nolint:ineffassign,staticcheck
```
Since `NewServerContext` returns a `context.Context` as well as a `ServerContext`, I'm overwriting the value of the passed in context (`ctx`) variable to reflect the most up to date value (the newly created child context).

The linter rejects this since the `ctx` variable isn't current used after this point.  I could satisfy the linter by assigning the returned context variable to `_`, but I think this invites the risk of anyone who updates this function in the future accidentally using the old value of `ctx` since they may not realize they need to update this line to get the correct context value.

awly replied:
Hmm, how about adding `ctx.Done()`  case to the select statement at the bottom of this method (in the for loop)?
And for `handleDirectTCPIPRequest` you could spawn a goroutine waiting on `ctx.Done()` to call `conn.Close()`
